src/equisolve/numpy/models/linear_model.py

```diff
@@ -0,0 +1,257 @@
+        # Convert problem with regularization term into an equivalent
+        # problem without the regularization term
+        num_targets, num_properties = X.shape
+        regularization_all = np.hstack((sample_weights, alphas))
+        regularization_eff = np.diag(np.sqrt(regularization_all))
+        X_eff = regularization_eff @ np.vstack((X, np.eye(num_properties)))
+        y_eff = regularization_eff @ np.hstack((y, np.zeros(num_properties)))
+        least_squares_output = np.linalg.lstsq(X_eff, y_eff, rcond=1e-13)
+        w_solver = least_squares_output[0]
+        return w_solver
+
+    def fit(
+        self, X: TensorMap, y: TensorMap, sample_weight: Optional[TensorMap] = None
+    ) -> None:
+        """Fit Ridge regression model to each block in X.
+
+        :param X: training data
+        :param y: target values
+        :param sample_weight: sample weights
+        """
+        self._validate_data(X, y)
+        self._validate_params(X, sample_weight)
+
+        for i_block, X_block in enumerate(X.blocks()):
+            y_block = y.block(i_block)
+            alpha_block = self.alpha.block(i_block)
+
+            # X_arr has shape of (n_targets, n_properties)
+            X_arr = block_to_array(X_block, self.parameter_keys)
+
+            # y_arr has shape lentgth of n_targets
+            y_arr = block_to_array(y_block, self.parameter_keys)[:, 0]
+
+            # alpha_arr has length of n_properties
+            alpha_arr = alpha_block.values[0]
+
+            # Sample weights
+            if sample_weight is not None:
+                sw_block = sample_weight.block(i_block)
+                # sw_arr has length of n_targets
+                sw_arr = block_to_array(sw_block, self.parameter_keys)[:, 0]
+                assert (
+                    sw_arr.shape == y_arr.shape
+                ), f"shapes = {sw_arr.shape} and {y_arr.shape}"
+            else:
+                sw_arr = np.ones((len(y_arr),))
+
+            w = self._numpy_lstsq_solver(X_arr, y_arr, sw_arr, alpha_arr)
+            self.coef_.append(w.reshape(-1, 1))
+
+        # Convert alpha to a dictionary to be used in external models.
+        self.alpha = tensor_map_to_dict(self.alpha)
+        return self
+
+    def predict(self, X: TensorMap) -> TensorMap:
+        """
+        Predict using the linear model.
+
+        :param X: samples
+        :returns: predicted values
+        """
+        # TODO general is fitted function
+        if len(self.coef_) == 0:
+            raise ValueError("No weights. Call fit method first.")
+
+        if len(X.blocks()) != len(self.coef_):
+            raise ValueError(
+                f"Number of blocks in X ({len(X.blocks())}) does not agree "
+                f"with the number of fitted weights ({len(self.coef_)})."
+            )
+        self.alpha = dict_to_tensor_map(self.alpha)
+
+        blocks = []
+        for i_block, X_block in enumerate(X.blocks()):
```

Comment:
```suggestion
        for key, X_block in X:
```

nd then correct `i_block` in the loop